libr/bin/format/xtac/bin_xtac.c

```diff
@@ -0,0 +1,587 @@
+		rbin->cb_printf ("    padding : 0x%x\n", blck_stub->padding);
+	}
+
+	rbin->cb_printf ("xtac linked list:\n");
+	RBinXtacLinkedListEntry *entry = NULL;
+	r_list_foreach (bin->xtac_linked_list, iter, entry) {
+		rbin->cb_printf ("  xtac linked list entry\n");
+		rbin->cb_printf ("    ptr_to_entry : 0x%x\n", entry->ptr_to_entry);
+		rbin->cb_printf ("    meta_data : 0x%x\n", GET_META_DATA (entry->meta_and_offset));
+		rbin->cb_printf ("    offset_to_next_entry : 0x%x\n", GET_OFFSET (entry->meta_and_offset) * 4);
+		rbin->cb_printf ("    forward_edge_addr : 0x%x\n", entry->forward_edge_addr);
+		rbin->cb_printf ("    backward_edge_addr : 0x%x\n", entry->backward_edge_addr);
+	}
+}
+
+static bool r_bin_xtac_read_header(RBinXtacObj *bin) {
+	if (!(bin->header = R_NEW0 (RBinXtacHeader))) {
+		r_sys_perror ("R_NEW0 (xtac header)");
+		return false;
+	}
+	if (r_buf_read_at (bin->b, 0, (ut8*) bin->header, sizeof (RBinXtacHeader)) < 0) {
+		R_LOG_WARN ("Warning: read (xtac header)\n");
+		free (bin->header);
+		bin->header = NULL;
+		return false;
+	}
+	return true;
+}
+
+static bool r_bin_xtac_read_address_pairs(RBinXtacObj *bin) {
+	const ut32 n_addr_pairs = bin->header->num_of_addr_pairs;
+	const ut32 addr_pair_size = n_addr_pairs * sizeof (X86ArmAddrPair);
+	const ut32 p_addr_pair = bin->header->ptr_to_addr_pairs;
+	if (!(bin->address_pairs = R_NEWS0 (X86ArmAddrPair, n_addr_pairs))) {
+		r_sys_perror ("R_NEWS0 (xtac address pair)");
+		return false;
+	}
+	if (r_buf_read_at (bin->b, p_addr_pair, (ut8*) bin->address_pairs, addr_pair_size) < 0) {
+		R_LOG_WARN ("Warning: read (xtac address pairs)\n");
+		free (bin->address_pairs);
+		bin->address_pairs = NULL;
+		return false;
+	}
+	return true;
+}
+
+static bool r_bin_xtac_read_module_name(RBinXtacObj *bin) {
+	const ut32 len_of_mod_name = bin->header->size_of_mod_name / sizeof (ut16) + 1;
+	const ut32 p_mod_name = bin->header->ptr_to_mod_name;
+	if (!(bin->mod_name_u16 = R_NEWS0 (ut16, len_of_mod_name))) {
+		r_sys_perror ("R_NEWS0 (xtac module name)");
+		return false;
+	}
+	if (r_buf_read_at (bin->b, p_mod_name, (ut8*) bin->mod_name_u16, bin->header->size_of_mod_name) < 0) {
+		R_LOG_WARN ("Warning: read (xtac module name)\n");
+		free (bin->mod_name_u16);
+		bin->mod_name_u16 = NULL;
+		return false;
+	}
+	if (!(bin->mod_name_u8 = r_str_utf16_decode((ut8*) bin->mod_name_u16, bin->header->size_of_mod_name))) {
+		free (bin->mod_name_u16);
+		bin->mod_name_u16 = NULL;
+		return false;
+	}
+	return true;
+}
+
+static bool r_bin_xtac_read_nt_native_pathname(RBinXtacObj *bin) {
+	const ut32 len_of_nt_pname = bin->header->size_of_nt_pname / sizeof (ut16) + 1;
+	const ut32 p_nt_name = bin->header->ptr_to_nt_pname;
+	if (!(bin->nt_path_name_u16 = R_NEWS0 (ut16, len_of_nt_pname))) {
+		r_sys_perror ("R_NEWS0 (xtac nt path)");
+		return false;
+	}
+	if (r_buf_read_at (bin->b, p_nt_name, (ut8*) bin->nt_path_name_u16, bin->header->size_of_nt_pname) < 0) {
```

Comment:
check for != size_of_nt_pname, this function will return the amount of read characters, not -1.

feel free to split that into two lines for readability purposes